README.md
# Duration Pretty

[![Build Status](https://travis-ci.org/Flcwl/duration-pretty.svg?branch=master)](https://travis-ci.org/github/Flcwl/duration-pretty)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/Flcwl/duration-pretty/blob/master/LICENSE)
[![npm version](https://img.shields.io/npm/v/duration-pretty.svg?style=flat)](https://www.npmjs.com/package/duration-pretty)

> The plugin is a pure JavaScript library that parses duration time length to format display.

## Getting Started

### Installation

```console
npm install --save duration-pretty
```

### Documentation

The `duration(timestamp, type)` get two parameters: timestamp && type.

```js
timestamp: number
type: seconds | milliseconds
```

Format duration time using a template string to `format()`.

```js
var duration = require('duration-pretty')

duration(7380, 'seconds').format('H:mm') // "2:03"
```

The template string is parsed for universal token characters, which are replaced with the duration's value for each unit type. The tokens are:

```js
days: D | DD
hours: H | HH
minutes: m | mm
seconds： s | ss
milliseconds: SSS
```

Escape token characters within the template string using `[]`.

```js
duration(3661, 'seconds').format('H [hrs], m [mins]') // "2 hrs, 3 mins"
```

## Tests

You can find all cases in `files:/test/*.spec.js`, And testing Using below script.

```console
npm run test
```


CHANGELOG.md
# Change Log

## v0.0.2

- [chore] Update README.md.
- [chore] separate the type as Enum `seconds` | `milliseconds`.
- [chore] update export as default

## v0.0.1

- [feat] Init duration-format.
- [feat] Add test.


test/duration-format.spec.js
import duration from '../src'

// $N means Nth parameter
describe('Time duration to format', () => {
  describe('milliseconds', () => {
    it('HH:mm:ss when $1 = 999', () => {
      const formattedDurationTime = duration(999, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:00')
    })

    it('HH:mm:ss when $1 = 1000', () => {
      const formattedDurationTime = duration(1000, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:01')
    })

    it('HH:mm:ss when $1 = 222222', () => {
      const formattedDurationTime = duration(222222, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:03:42')
    })

    it('HH:mm:ss when $1 = 2222', () => {
      const formattedDurationTime = duration(2222, 'milliseconds').format('HH:mm:ss:SSS')
      expect(formattedDurationTime).toEqual('00:00:02:222')
    })

    it('HH:mm:ss when $1 = 273222222', () => {
      const formattedDurationTime = duration(273222222, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('75:53:42')
    })

    it('HH,mm[[H]]ss when $1 = 12732222', () => {
      const formattedDurationTime = duration(12732222, 'milliseconds').format('HH,mm[[H]]ss')
      expect(formattedDurationTime).toEqual('03,32[H]12')
    })

    it('HH:mm:ss when $1 = -1000', () => {
      const formattedDurationTime = duration(-1000, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:00')
    })

    it('HH:mm:ss when $1 = -12732222', () => {
      const formattedDurationTime = duration(-12732222, 'milliseconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:00')
    })
  })

  describe('seconds', () => {
    it('HH:mm:ss when $1 = 1', () => {
      const formattedDurationTime = duration(1, 'seconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:01')
    })

    it('HH:mm:ss when $1 = 1000', () => {
      const formattedDurationTime = duration(1000, 'seconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:16:40')
    })

    it('HH:mm:ss when $1 = 7380', () => {
      const formattedDurationTime = duration(7380, 'seconds').format('H:mm')
      expect(formattedDurationTime).toEqual('2:03')
    })

    it('HH:mm:ss when $1 = -1', () => {
      const formattedDurationTime = duration(-1, 'seconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:00')
    })

    it('HH:mm:ss when $1 = -12732222', () => {
      const formattedDurationTime = duration(-12732222, 'seconds').format('HH:mm:ss')
      expect(formattedDurationTime).toEqual('00:00:00')
    })

    it('HH:mm:ss when $1 = 3661', () => {
      const formattedDurationTime = duration(3661, 'seconds').format('H [hrs], m [mins]')
      expect(formattedDurationTime).toEqual('1 hrs, 1 mins')
    })
  })
})


src/duration-format.ts
import Utils from './utils'

interface DurationTime {
  days: number
  hours: number
  minutes: number
  seconds: number
  milliseconds: number
}

enum DurationType {
  seconds = 'seconds',
  milliseconds = 'milliseconds',
}

interface Duration {
  $d: DurationTime
  $t: number
  format: (formatStr: string) => string
}

const MS_PER_SECOND = 1e3
const SECONDS_PER_MIN = 60
const MINS_PER_HOUR = 60
const HOURS_PER_DAY = 24
const DAYS_PER_MONTH = 30
const DAYS_PER_YEAR = 365

const ONE_MS = 1
const ONE_SECOND = ONE_MS * MS_PER_SECOND // second
const ONE_MIN = ONE_SECOND * SECONDS_PER_MIN // minute
const ONE_HOUR = ONE_MIN * MINS_PER_HOUR // hour
const ONE_DAY = ONE_HOUR * HOURS_PER_DAY // day
const MONTH = ONE_DAY * DAYS_PER_MONTH // month
const ONE_YEAR = ONE_DAY * DAYS_PER_YEAR // year

const DAYS = 0
const HOURS = 1
const MINUTES = 2
const SECONDS = 3
const MILLISECONDS = 4
const RULES_LENGTH = 5
const matchRules = [
  { label: 'D', name: 'days', pos: DAYS, next: HOURS_PER_DAY },
  { label: 'H', name: 'hours', pos: HOURS, next: MINS_PER_HOUR },
  { label: 'm', name: 'minutes', pos: MINUTES, next: SECONDS_PER_MIN },
  { label: 's', name: 'seconds', pos: SECONDS, next: MS_PER_SECOND },
  { label: 'SSS', name: 'milliseconds', pos: MILLISECONDS, next: ONE_MS },
]
const REGEX_FORMAT = /\[([^\]]+)]|D{1,2}|H{1,2}|m{1,2}|s{1,2}|SSS/g
const TIME_TRANSFORM = {
  toSeconds(time: number) {
    return Math.floor(time / ONE_SECOND)
  },

  toMinutes(time: number) {
    return Math.floor(time / ONE_MIN)
  },

  toHours(time: number) {
    return Math.floor(time / ONE_HOUR)
  },

  toDays(time: number) {
    return Math.floor(time / ONE_DAY)
  },

  toMonths(time: number) {
    return Math.floor(time / MONTH)
  },

  toYears(time: number) {
    return Math.floor(time / ONE_YEAR)
  },

  milliseconds(ms: number): DurationTime {
    const days = this.toDays(ms)
    const hours = this.toHours(ms) % HOURS_PER_DAY
    const minutes = this.toMinutes(ms) % MINS_PER_HOUR
    const seconds = this.toSeconds(ms) % SECONDS_PER_MIN
    const milliseconds = Math.floor(ms) % MS_PER_SECOND
    return {
      days,
      hours,
      minutes,
      seconds,
      milliseconds,
    }
  },

  seconds(seconds: number): DurationTime {
    return this.milliseconds(seconds * 1000)
  },
}

export const duration = (t: unknown, type: DurationType = DurationType.milliseconds): Duration => {
  if (typeof t !== 'number') {
    throw new TypeError('Expected a number')
  }

  const formatFunc = TIME_TRANSFORM[type]

  if (!formatFunc) {
    throw new TypeError('Expected a string in milliseconds|seconds')
  }

  const dur = formatFunc.call(TIME_TRANSFORM, Math.abs(t < 0 ? 0 : t))

  return {
    $d: dur,
    $t: t,
    format(formatStr = 'HH:mm:ss-SSS') {
      const visited = Utils.l(RULES_LENGTH, false)
      const dates = Utils.l(RULES_LENGTH, 0)
      let t = 0

      matchRules.forEach((item) => {
        if (!formatStr.match(item.label)) {
          t += dur[item.name] * item.next
          return
        }

        if (!visited[item.pos]) {
          dates[item.pos] = t + dur[item.name]
          t = 0
          visited[item.pos] = true
        }
      })

      const matches = {
        D: dates[DAYS],
        DD: Utils.z(dates[DAYS]),
        H: String(dates[HOURS]),
        HH: Utils.z(dates[HOURS]),
        m: String(dates[MINUTES]),
        mm: Utils.z(dates[MINUTES]),
        s: String(dates[SECONDS]),
        ss: Utils.z(dates[SECONDS]),
        SSS: Utils.z(dates[MILLISECONDS], 3),
      }

      // replace $1
      return formatStr.replace(REGEX_FORMAT, (match, $1) => $1 || matches[match])
    },
  }
}


lib/duration-format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.duration = void 0;
const utils_1 = require("./utils");
var DurationType;
(function (DurationType) {
    DurationType["seconds"] = "seconds";
    DurationType["milliseconds"] = "milliseconds";
})(DurationType || (DurationType = {}));
const MS_PER_SECOND = 1e3;
const SECONDS_PER_MIN = 60;
const MINS_PER_HOUR = 60;
const HOURS_PER_DAY = 24;
const DAYS_PER_MONTH = 30;
const DAYS_PER_YEAR = 365;
const ONE_MS = 1;
const ONE_SECOND = ONE_MS * MS_PER_SECOND; // second
const ONE_MIN = ONE_SECOND * SECONDS_PER_MIN; // minute
const ONE_HOUR = ONE_MIN * MINS_PER_HOUR; // hour
const ONE_DAY = ONE_HOUR * HOURS_PER_DAY; // day
const MONTH = ONE_DAY * DAYS_PER_MONTH; // month
const ONE_YEAR = ONE_DAY * DAYS_PER_YEAR; // year
const DAYS = 0;
const HOURS = 1;
const MINUTES = 2;
const SECONDS = 3;
const MILLISECONDS = 4;
const RULES_LENGTH = 5;
const matchRules = [
    { label: 'D', name: 'days', pos: DAYS, next: HOURS_PER_DAY },
    { label: 'H', name: 'hours', pos: HOURS, next: MINS_PER_HOUR },
    { label: 'm', name: 'minutes', pos: MINUTES, next: SECONDS_PER_MIN },
    { label: 's', name: 'seconds', pos: SECONDS, next: MS_PER_SECOND },
    { label: 'SSS', name: 'milliseconds', pos: MILLISECONDS, next: ONE_MS },
];
const REGEX_FORMAT = /\[([^\]]+)]|D{1,2}|H{1,2}|m{1,2}|s{1,2}|SSS/g;
const TIME_TRANSFORM = {
    toSeconds(time) {
        return Math.floor(time / ONE_SECOND);
    },
    toMinutes(time) {
        return Math.floor(time / ONE_MIN);
    },
    toHours(time) {
        return Math.floor(time / ONE_HOUR);
    },
    toDays(time) {
        return Math.floor(time / ONE_DAY);
    },
    toMonths(time) {
        return Math.floor(time / MONTH);
    },
    toYears(time) {
        return Math.floor(time / ONE_YEAR);
    },
    milliseconds(ms) {
        const days = this.toDays(ms);
        const hours = this.toHours(ms) % HOURS_PER_DAY;
        const minutes = this.toMinutes(ms) % MINS_PER_HOUR;
        const seconds = this.toSeconds(ms) % SECONDS_PER_MIN;
        const milliseconds = Math.floor(ms) % MS_PER_SECOND;
        return {
            days,
            hours,
            minutes,
            seconds,
            milliseconds,
        };
    },
    seconds(seconds) {
        return this.milliseconds(seconds * 1000);
    },
};
exports.duration = (t, type = DurationType.milliseconds) => {
    if (typeof t !== 'number') {
        throw new TypeError('Expected a number');
    }
    const formatFunc = TIME_TRANSFORM[type];
    if (!formatFunc) {
        throw new TypeError('Expected a string in milliseconds|seconds');
    }
    const dur = formatFunc.call(TIME_TRANSFORM, Math.abs(t < 0 ? 0 : t));
    return {
        $d: dur,
        $t: t,
        format(formatStr = 'HH:mm:ss-SSS') {
            const visited = utils_1.default.l(RULES_LENGTH, false);
            const dates = utils_1.default.l(RULES_LENGTH, 0);
            let t = 0;
            matchRules.forEach((item) => {
                if (!formatStr.match(item.label)) {
                    t += dur[item.name] * item.next;
                    return;
                }
                if (!visited[item.pos]) {
                    dates[item.pos] = t + dur[item.name];
                    t = 0;
                    visited[item.pos] = true;
                }
            });
            const matches = {
                D: dates[DAYS],
                DD: utils_1.default.z(dates[DAYS]),
                H: String(dates[HOURS]),
                HH: utils_1.default.z(dates[HOURS]),
                m: String(dates[MINUTES]),
                mm: utils_1.default.z(dates[MINUTES]),
                s: String(dates[SECONDS]),
                ss: utils_1.default.z(dates[SECONDS]),
                SSS: utils_1.default.z(dates[MILLISECONDS], 3),
            };
            // replace $1
            return formatStr.replace(REGEX_FORMAT, (match, $1) => $1 || matches[match]);
        },
    };
};
